auth-passwd.c

```diff
@@ -298,9 +298,24 @@ sys_auth_passwd(struct ssh *ssh, const char *password)
 	wchar_t * backslash = wcschr(user_utf16, L'\\');
 	if (backslash != NULL) {
 
+		/* assemble the path to the name translation library */
+		wchar_t library_path[MAX_PATH + 1];
+		if (GetSystemDirectoryW(library_path, ARRAYSIZE(library_path)) == 0) {
+			debug3("%s: GetSystemDirectoryW() failed name translation: %d", __FUNCTION__, GetLastError());
+			goto done;
+		}
+		wcscat_s(library_path, ARRAYSIZE(library_path), L"\\secur32.dll");
+
+		/* dynamically load name translation function to support static linking to onecore */
+		typedef (*TranslateNameWFunc)(_In_ LPCWSTR lpAccountName,
+			_In_ EXTENDED_NAME_FORMAT AccountNameFormat, _In_ EXTENDED_NAME_FORMAT DesiredNameFormat,
+			_Out_writes_to_opt_(*nSize, *nSize) LPWSTR lpTranslatedName, _Inout_ PULONG nSize);
+		HMODULE library = LoadLibraryW(library_path);
+		TranslateNameWFunc LocalTranslateNameW = NULL;
+
 		/* attempt to format into upn format as this is preferred for login */
-		if (TranslateNameW(user_utf16, NameSamCompatible, 
-			NameUserPrincipal, domain_upn, &domain_upn_len) != 0) {
+		if (library != NULL && (LocalTranslateNameW = (TranslateNameWFunc) GetProcAddress(library, "TranslateNameW")) != NULL &&
+			LocalTranslateNameW(user_utf16, NameSamCompatible, NameUserPrincipal, domain_upn, &domain_upn_len) != 0) {
```

Comment:
Does this work?

wchar_t * user_name = user_utf16;
	wchar_t * domain_name = NULL;
	wchar_t * seperator = wcschr(user_name, L'\\');
	if (seperator != NULL) {
		domain_name = user_name;
		*seperator = L'\0';
		user_name = seperator + 1;
	}

Reply:
Yes and no.  That's already the 'else' condition.  The intent here was to prefer using LogonUserExExWHelper(UserPrincipalName, NULL, ...) over LogonUserExExWHelper(SamAccoutName, NetBiosDomainName, ...).  There have been some cases where UserPrincipalName (username@domain.local.com) has purportedly been a better format to use for reliable logon.  UserPrincipalName is definitely important for Kerberos S4U, but I've never _personally_ seen an advantage to using it for logon.  I would gladly remove this if we don't think it's value-added here.

Reply:
I didn't debug to the code. in the return from get_passwd. Do you know even with sam format, does it contain the domain fqdn name or only the domain netbios name?

Reply:
It will contain SamAccountName for local accounts or NetBiosDomain\SamAccountName for domain accounts.  Even if it did return DomainFqdn\SamAccountName, the UserPrincipalName domain suffix can often be something other than the DomainFqdn.  This is especially common where in forests with child domains in which can the child domain accounts can have the UserPrincipalName suffix of the parent domain.  So, unfortunately, the only reliable way to translate NetBiosDomain\SamAccountName or DomainFqdn\SamAccountName into UserPrincipalName is to query Active Directory which is what TranslateNameW() does under the credentials of the service account.